kong/tools/string.lua

```diff
@@ -0,0 +1,45 @@
+local find          = string.find
+local gsub          = string.gsub
+
+
+local _M = {}
+
+
+local replace_dashes
+local replace_dashes_lower
+do
+  local str_replace_char
+
+  if ngx and ngx.config.subsystem == "http" then
+
+    -- 1,000,000 iterations with input of "my-header":
+    -- string.gsub:        81ms
+    -- ngx.re.gsub:        74ms
+    -- loop/string.buffer: 28ms
+    -- str_replace_char:   14ms
+    str_replace_char = require("resty.core.utils").str_replace_char
+
+  else    -- stream subsystem
+    str_replace_char = function(str, ch, replace)
+      if not find(str, ch, nil, true) then
+        return str
+      end
+
+      return gsub(str, ch, replace)
+    end
+  end
+
+  replace_dashes = function(str)
```

Comment:
hi @zhongweiy if we wrap a function with multiple function calls, could luajit optimize the calling chain to directly call the actual function?

I mean if calling replaces_dashes() and calling str_replace_char() has similar performance?

Reply:
with a simple test, I found that wrapping a function does not add extra time overhead.

Reply:
@chobits yeah, luajit can do this kind of optimization when JIT compilation is triggered (it is called inlining). And with the default configuration, inline level is 3. It means 3 level of function call will be inlined. 

But our case is a tail call, tail call has more optimization (the inline can happen regardless the call level).

Reply:
> @chobits yeah, luajit can do this kind of optimization when JIT compilation is triggered (it is called inlining). And with the default configuration, inline level is 3. It means 3 level of function call will be inlined.
> 
> But our case is a tail call, tail call has more optimization (the inline can happen regardless the call level).

Pretty cool. Thanks for the details.